fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OlapTablePredicateExtractor.java

```diff
@@ -0,0 +1,158 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rewrite;
+
+import com.starrocks.catalog.Column;
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.LambdaFunctionOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+// extract olap table scan predicates to two parts: pushed-down to scan node, keep into filter node
+// extract the predicates of table scan and divide it into two part:
+// 1. pushedPredicates: simple single-column predicates that can use the optimizations of storage layer,
+// these should be evaluated inside ScanNode;
+// 2. reservedPredicates: other predicates which can be evaluated outside ScanNode;
+public class OlapTablePredicateExtractor {
```

Comment:
Can this be used for external tables too later?

Reply:
> Can this be used for external tables too later?

in fact it has been used for external tables other than mysql/jdbc/odps external tables.  the original `OlapTablePredicateExtractor` is not exactly, I changed it to `TableScanPredicateExtractor`. we can implement it seperately for each type of tables if there is a specific need.